core/shared/src/main/scala/scalaz/zio/ZIO.scala

```diff
@@ -1078,219 +1078,219 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+   * because a defect has been detected in the code.
    */
-  final def supervise[R >: LowerR, E <: UpperE, A](zio: ZIO[R, E, A]): ZIO[R, E, A] =
-    superviseWith(zio)(Fiber.interruptAll)
+  final def dieMessage(message: String): UIO[Nothing] = die(new RuntimeException(message))
 
   /**
-   * Returns an effect that supervises the specified effect, ensuring that all
-   * fibers that it forks are passed to the specified supervisor as soon as the
-   * supervised effect completes.
+   * Returns an effect from a [[scalaz.zio.Exit]] value.
    */
-  final def superviseWith[R >: LowerR, E <: UpperE, A](
-    zio: ZIO[R, E, A]
-  )(supervisor: IndexedSeq[Fiber[_, _]] => ZIO[R, Nothing, _]): ZIO[R, E, A] =
-    zio.ensuring(children.flatMap(supervisor(_))).supervised
+  final def done[E, A](r: Exit[E, A]): IO[E, A] = r match {
+    case Exit.Success(b)     => succeed(b)
+    case Exit.Failure(cause) => halt(cause)
+  }
 
   /**
-   * Returns an effect that first executes the outer effect, and then executes
-   * the inner effect, returning the value from the inner effect, and effectively
-   * flattening a nested effect.
+   * Returns information about the current fiber, such as its identity.
    */
-  final def flatten[R >: LowerR, E <: UpperE, A](zio: ZIO[R, E, ZIO[R, E, A]]): ZIO[R, E, A] =
-    zio.flatMap(ZIO.identityFn[ZIO[R, E, A]])
+  final def descriptor: UIO[Fiber.Descriptor] = descriptorWith(succeed(_))
 
   /**
-   * Returns a lazily constructed effect, whose construction may itself require
-   * effects. This is a shortcut for `flatten(effectTotal(io))`.
+   * Constructs an effect based on information about the current fiber, such as
+   * its identity.
    */
-  final def suspend[R >: LowerR, E <: UpperE, A](io: => ZIO[R, E, A]): ZIO[R, E, A] =
-    flatten(effectTotal(io))
+  final def descriptorWith[R, E, A](f: Fiber.Descriptor => ZIO[R, E, A]): ZIO[R, E, A] =
+    new ZIO.Descriptor(f)
 
   /**
-   * Returns an effect that will execute the specified effect fully on the
-   * provided executor, before returning to the default executor.
+   *
+   * Imports a synchronous effect into a pure `ZIO` value, translating any
+   * throwables into a `Throwable` failure in the returned value.
+   *
+   * {{{
+   * def putStrLn(line: String): Task[Unit] = Task.effect(println(line))
+   * }}}
    */
-  final def lock[R >: LowerR, E <: UpperE, A](executor: Executor)(zio: ZIO[R, E, A]): ZIO[R, E, A] =
-    new ZIO.Lock(executor, zio)
+  final def effect[A](effect: => A): Task[A] = new ZIO.EffectPartial(() => effect)
 
   /**
    * Imports an asynchronous effect into a pure `ZIO` value. See `effectAsyncMaybe` for
    * the more expressive variant of this function that can return a value
    * synchronously.
    */
-  final def effectAsync[E <: UpperE, A](register: (ZIO[Any, E, A] => Unit) => Unit): ZIO[Any, E, A] =
+  final def effectAsync[E, A](register: (ZIO[Any, E, A] => Unit) => Unit): ZIO[Any, E, A] =
```

Comment:
Thanks to @iravid's change, I think this can now use `R`. Be sure to copy over his changes and any other new ones when you pull in latest changes from master.